src/writer/creator.cpp

```diff
@@ -265,13 +273,16 @@ namespace zim
 
       TINFO("Waiting for workers");
       // wait all cluster compression has been done
+      // If we are in error state, threads have been stoped and waiting_task
+      // will never reach 0, so no need to wait.
       unsigned int wait = 0;
       do {
         microsleep(wait);
         wait += 10;
-      } while(ClusterTask::waiting_task.load() > 0);
+      } while(ClusterTask::waiting_task.load() > 0 && !data->isErrored());
```

Comment:
This is duplication with `XapianHandler::waitNoMoreTask()`.  There is also some duplication across `IndexTask` and `ClusterTask` due to maintenance of the waiting task count. All code duplication can be eliminated via an introduction of an auxiliary template class.

Draft

```c++
template<class T>
class TrackableTask : public Task {
public:
    TrackableTask(const TrackableTask&) = delete;
    TrackableTask& operator=(const TrackableTask&) = delete;
    TrackableTask() { ++waitingTaskCount; }
    virtual ~TrackableTask() { --waitingTaskCount; }

    void waitNoMoreTask(const CreatorData* data)
    {
        unsigned int wait = 0;
        do {
            microsleep(wait);
            wait += 10;
        } while(waitingTaskCount.load() > 0 && !data->isErrored());
    }

private:
    static std::atomic<unsigned long> waitingTaskCount;
};

class ClusterTask : TrackableTask<ClusterTask>
{
    ...
};

class IndexTask : TrackableTask<IndexTask>
{
    ... 
};
```

Reply:
Indeed, good idea.